File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p02895/p02895_num2_parsed.py
from itertools import permutations

def is_valid(pairs, matrix):
    for i, j in pairs:
        if matrix[i][j] != 1 or matrix[j][i] != 1:
            return False
    return True

def count_ways(N, A):
    # Convert input string to boolean matrix
    matrix = [[bool(int(A[i][j])) for j in range(2*N)] for i in range(2*N)]
    
    # Generate all possible pairs of 2N points and check validity
    all_pairs = list(permutations(range(2*N), 2*N))
    valid_pairs = []
    
    for pairs in all_pairs:
        if is_valid(pairs, matrix):
            valid_pairs.append(pairs)
            
    # Count the number of valid ways to divide the 2N points into N pairs
    count = 0
    
    for pairs in valid_pairs:
        visited = set()
        components = 0
        
        for i in range(2*N):
            if i not in visited:
                stack = [i]
                components += 1
                
                while stack:
                    node = stack.pop()
                    visited.add(node)
                        
                    for j in range(2*N):
                        if j not in visited and (node, j) in pairs:
                            stack.append(j)
                            
        if components == N:
            count += 1
            
    return count


# Read input
N = int(input())
A = [input() for _ in range(2*N)]

# Count the number of ways to divide the 2N points into N pairs
result = count_ways(N, A)

# Print the result
print(result)

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p02895/p02895_num2_parsed.cpp
#include <iostream>
#include <vector>
using namespace std;

long long countWays(int N, vector<vector<int>>& A, vector<vector<long long>>& dp, int mask) {
    // Base case: If all pairs are formed, return 1
    if (mask == (1 << (2 * N)) - 1) {
        return 1;
    }
    
    // If dp state is already calculated, return the value
    if (dp[mask][N] != -1) {
        return dp[mask][N];
    }
    
    long long ways = 0;
    
    // Find the first point that is not paired yet
    int i = 0;
    while ((mask & (1 << i)) != 0) {
        i++;
    }
    
    // Pair the point with all other unpaired points satisfying the conditions
    for (int j = i + 1; j < 2 * N; j++) {
        if (((mask & (1 << j)) == 0) && A[i][j] == 1) {
            ways += countWays(N - 1, A, dp, mask | (1 << i) | (1 << j));
        }
    }
    
    // Memorize the calculated value in dp table
    dp[mask][N] = ways;
    
    return ways;
}

int main() {
    // Read input
    int N;
    cin >> N;
    
    // Initialize A matrix
    vector<vector<int>> A(2 * N, vector<int>(2 * N));
    for (int i = 0; i < 2 * N; i++) {
        string row;
        cin >> row;
        for (int j = 0; j < 2 * N; j++) {
            A[i][j] = row[j] - '0';
        }
    }
    
    // Initialize dp table
    vector<vector<long long>> dp(1 << (2 * N), vector<long long>(N + 1, -1));
    
    // Count the number of ways to divide the points into pairs
    long long ways = countWays(N, A, dp, 0);
    
    // Print the result
    cout << ways << endl;
    
    return 0;
}